---
title: "Skin UVB SKH1 mouse model treated with SFN "
output:
  html_notebook:
    toc: yes
    toc_float: yes
    code_folding: hide
---

# Part 1: RNA
```{r header, echo = FALSE, message = FALSE, error = FALSE, warning  =FALSE}
# if (!requireNamespace("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
# BiocManager::install("DESeq2")

require(knitr)
require(data.table)
require(DT)
require(DESeq2)
require(readxl)
require(BiocParallel)
require(ggplot2)
require(plotly)
require(threejs)
require(scales)

# NOTE: on DESeq2 Output: 'baseMean' is the average of the normalized count values, 
# divided by the size factors, taken over all samples in the DESeqDataSet
```

## Load RNA samples
Out of 30 samples, we selected 18 for this study. These are the normal tissue samples form the control, the UVA and the UVA+SFN treatment groups. normal tissue samples from the UVB_UA groups as well as tumor samples were excluded from this analysis.     
First, we removed 7,148 genes with zero counts in > 80% (> 14 out of 18) of samples. 17,273 out of 24,421 genes left. 
         
```{r data_rna, warning = FALSE, echo = FALSE, message = FALSE}
# Load data----
dt0 <- fread("data/renyi_dedup_rnaseq_data/featurescounts_uvb-skin_dedup_renyi_2-9-2018.csv",
             skip = 1)

# Remove unused columns----
dt1 <- dt0[, c(1, 6:ncol(dt0)), with = FALSE]

cnames <- colnames(dt1)[-c(1:2)]
cnames <- gsub(x = cnames,
               pattern = ".dedup.bam",
               replacement = "")
colnames(dt1)[-c(1:2)] <- cnames

# ATTENTION! In this analysis, we will only examine controls and SFN
# Also, removed cancer cell samples
tnames <- substr(x = colnames(dt1), 
                 start = 3,
                 stop = 3)

gnames <- substr(x = colnames(dt1), 
                 start = 5,
                 stop = 7)

dt1 <- dt1[, gnames %in% c("id",
                           "th",
                           "CON",
                           "UVB",
                           "SFN" ) &
             tnames != "t",
           with = FALSE]
# 18 samples left

# Remove genes with zero counts in > 80% (> 14 out of 18) of samples
tmp <- dt1[, -c(1:2)] == 0
tmp <- rowSums(tmp) > 14
sum(tmp)

dt1 <- droplevels(dt1[!tmp, ])
nrow(dt1)
# 17,273 out of 24,421 genes left

datatable(head(dt1, 10),
              rownames = FALSE,
              options = list(pageLength = 10),
              caption = "Table 1: first 10 rows of the count table")
```

## Transcripts per kilobase million (TPM) normalization
Next, we noramized the counts. To convert number of hits to  the relative abundane of genes in each sample, we used ***transcripts per kilobase million (TPM)*** normalization, which is as following for the j-th sample:       
1. normilize for gene length: a[i, j] = 1,000*count[i, j]/gene[i, j] length(bp)     
2. normalize for seq depth (i.e. total count): a(i, j)/sum(a[, j])     
3. multiply by one million     
A very good comparison of normalization techniques can be found at the following video:    
[RPKM, FPKM and TPM, clearly explained](https://www.rna-seqblog.com/rpkm-fpkm-and-tpm-clearly-explained/)
     
After the normalization, each sample's total is 1M:
     
```{r tpm, warning = FALSE, echo = FALSE, message = FALSE}
# Normalize counts to TPM
tmp <- 1000*dt1[, 3:ncol(dt1)]/dt1$Length
tpm <- data.table(Geneid = dt1$Geneid,
                  Length = dt1$Length,
                  apply(tmp,
                        2,
                        function(a) {
                          10^6*(a/sum(a))
                        }))
colSums(tpm[, -c(1:2)])

formatRound(datatable(head(tpm, 10),
                      rownames = FALSE,
                      options = list(pageLength = 10),
                      caption = "Table 2: transcripts per kilobase million (TPM) normalized counts"),
            columns = 3:ncol(tpm),
            digits = 2)

# Total TPM
total <- rowSums(tpm[, 3:ncol(tpm)])

# Sort genes by relative abundancy
tpm$Geneid <- factor(tpm$Geneid ,
                     levels = tpm$Geneid[order(total,
                                               decreasing = FALSE)])
```

# Top 100 most abundant RNA molecules
```{r most_abundant}
# Separate top 100 abundant genes
tmp <- droplevels(tpm[Geneid %in% levels(tpm$Geneid)[(nrow(tpm) - 99):nrow(tpm)]])

tmp <- melt.data.table(data = tmp,
                       id.vars = 1:2,
                       measure.vars = 3:ncol(tmp),
                       variable.name = "Sample",
                       value.name = "TPM")

tmp$Week <- substr(x = tmp$Sample,
                   start = 1,
                   stop = 3)
tmp$Week <- factor(tmp$Week,
                   levels = unique(tmp$Week))


tmp$Treatment <- substr(x = tmp$Sample,
                        start = 5,
                        stop = 7)
tmp$Treatment <- factor(tmp$Treatment,
                        levels = c("CON", 
                                   "UVB",
                                   "SFN"))

tmp$Replica <- substr(x = tmp$Sample,
                      start = 9,
                      stop = 9)
tmp$Replica <- factor(tmp$Replica,
                      levels = 0:1)

# Plot top 100 abundant genes
p2 <- ggplot(tmp,
             aes(x = TPM,
                 y = Geneid,
                 fill = Treatment,
                 shape = Week)) +
  # facet_wrap(~ Sex, nrow = 1) +
  geom_point(size = 3,
             alpha = 0.5) +
  geom_vline(xintercept = 1,
             linetype = "dashed")
ggplotly(p2)
```

# Bottom 100 least abundant RNA molecules
```{r least_abundant}
tmp <- droplevels(tpm[Geneid %in% levels(tpm$Geneid)[1:100]])

tmp <- melt.data.table(data = tmp,
                       id.vars = 1:2,
                       measure.vars = 3:ncol(tmp),
                       variable.name = "Sample",
                       value.name = "TPM")

tmp$Week <- substr(x = tmp$Sample,
                   start = 1,
                   stop = 3)
tmp$Week <- factor(tmp$Week,
                   levels = unique(tmp$Week))


tmp$Treatment <- substr(x = tmp$Sample,
                        start = 5,
                        stop = 7)
tmp$Treatment <- factor(tmp$Treatment,
                        levels = c("CON", 
                                   "UVB",
                                   "SFN"))

tmp$Replica <- substr(x = tmp$Sample,
                      start = 9,
                      stop = 9)
tmp$Replica <- factor(tmp$Replica,
                      levels = 0:1)

# Plot top 100 abundant genes
p3 <- ggplot(tmp,
             aes(x = TPM,
                 y = Geneid,
                 fill = Treatment,
                 shape = Week)) +
  # facet_wrap(~ Sex, nrow = 1) +
  geom_point(size = 3,
             alpha = 0.5) +
  geom_vline(xintercept = 1,
             linetype = "dashed")
ggplotly(p3)
```

# Meta data
```{r meta}
dmeta <- data.table(Sample = colnames(dt1)[-c(1:2)])

dmeta$time <- substr(x = dmeta$Sample,
                   start = 1,
                   stop = 3)
dmeta$time <- factor(dmeta$time,
                   levels = c("02w",
                              "15w",
                              "25w"))
dmeta$Week <- factor(dmeta$time,
                   levels = c("02w",
                              "15w",
                              "25w"),
                   labels = c("Week 2",
                              "Week 15",
                              "Week 25"))

dmeta$trt <- substr(x = dmeta$Sample,
                        start = 5,
                        stop = 7)
dmeta$trt <- factor(dmeta$trt,
                        levels = c("CON", 
                                   "UVB",
                                   "SFN"))
dmeta$Treatment <- factor(dmeta$trt,
                        levels = c("CON", 
                                   "UVB",
                                   "SFN"),
                        labels = c("Negative Control",
                                   "Positive Control (UVB)",
                                   "Sulforaphane (SFN)"))

dmeta$Replica <- substr(x = dmeta$Sample,
                      start = 9,
                      stop = 9)
dmeta$Replica <- factor(dmeta$Replica,
                      levels = 0:1)

datatable(dmeta,
          options = list(pageLength = nrow(dmeta)))
```

# PCA of TPM
NOTE: the distributions are skewed. To make them symmetric, log transformation is often applied. However, there is an issue of zeros. In this instance, we added a small values ***lambda[i]*** equal to 1/10 of the smallest non-zero value of *i*-th gene. 
```{r pca}
dm.tpm <- as.matrix(tpm[, -c(1:2), with = FALSE])
rownames(dm.tpm) <- tpm$Geneid

# # Remove 02w_CON_1 sample and redo PCA
# dm.tpm <- dm.tpm[, colnames(dm.tpm) != "02w_CON_1"]
# dmeta <- dmeta[dmeta$Sample != "02w_CON_1", ]

# Add lambdas to all values, then take a log
dm.ltpm <- t(apply(X = dm.tpm,
                      MARGIN = 1,
                      FUN = function(a) {
                        lambda <- min(a[a > 0])/10
                        log(a + lambda)
                      }))

# PCA----
m1 <- prcomp(t(dm.ltpm),
             center = TRUE,
             scale. = TRUE)

s1 <- summary(m1)
s1
```

# Pareto chart of variance explained by principal components
```{r pca_var_plot}
imp <- data.table(PC = colnames(s1$importance),
                  Variance = 100*s1$importance[2, ],
                  Cumulative = 100*s1$importance[3, ])
imp$PC <- factor(imp$PC,
                 levels = imp$PC)
p1 <- ggplot(imp,
             aes(x = PC,
                 y = Variance)) +
  geom_bar(stat = "identity",
           fill = "grey",
           color = "black") +
  geom_line(aes(y = rescale(Cumulative,
                            to = c(min(Cumulative)*30/100,
                                   30)),
                group = rep(1, nrow(imp)))) +
  geom_point(aes(y = rescale(Cumulative,
                             to = c(min(Cumulative)*30/100,
                                    30)))) +
  scale_y_continuous("% Variance Explained",
                     breaks = seq(0, 30, by = 5),
                     labels = paste(seq(0, 30, by = 5),
                                    "%",
                                    sep = ""),
                     sec.axis = sec_axis(trans = ~.,
                                         name = "% Cumulative Variance",
                                         breaks = seq(0, 30, length.out = 5),
                                         labels = paste(seq(0, 100, length.out = 5),
                                                        "%",
                                                        sep = ""))) +
  scale_x_discrete("") +
  theme(axis.text.x = element_text(angle = 90,
                                   hjust = 1))
p1
```

# First 3 principal components, pairwise
```{r pca_plots}
# Biplot while keep only the most important variables (Javier)----
# Select PC-s to pliot (PC1 & PC2)
choices <- c(1:3)

# Scores, i.e. points (df.u)
dt.scr <- data.table(m1$x[, choices])
# Add grouping variables
dt.scr$trt <- dmeta$trt
dt.scr$time <- dmeta$time
dt.scr$sample <- dmeta$Sample

# Loadings, i.e. arrows (df.v)
dt.rot <- as.data.frame(m1$rotation[, choices])
dt.rot$feat <- rownames(dt.rot)
dt.rot <- data.table(dt.rot)

# Axis labels
u.axis.labs <- paste(colnames(dt.rot)[choices], 
                     sprintf('(%0.1f%% explained var.)', 
                             100*m1$sdev[choices]^2/sum(m1$sdev^2)))

p1 <- ggplot(data = dt.scr,
             aes(x = PC1,
                 y = PC2,
                 color = trt,
                 shape = time)) +
  geom_point(size = 4,
             alpha = 0.5) +
  scale_x_continuous(u.axis.labs[1]) +
  scale_y_continuous(u.axis.labs[2])
ggplotly(p1)

p1 <- ggplot(data = dt.scr,
             aes(x = PC1,
                 y = PC3,
                 color = trt,
                 shape = time)) +
  geom_point(size = 4,
             alpha = 0.5) +
  scale_x_continuous(u.axis.labs[1]) +
  scale_y_continuous(u.axis.labs[3])
ggplotly(p1)

p1 <- ggplot(data = dt.scr,
             aes(x = PC2,
                 y = PC3,
                 color = trt,
                 shape = time)) +
  geom_point(size = 4,
             alpha = 0.5) +
  scale_x_continuous(u.axis.labs[2]) +
  scale_y_continuous(u.axis.labs[3])
ggplotly(p1)
```

# First 3 principal components, 3D
```{r pca_3d, fig.height = 10, fig.width = 10}
scatterplot3js(x = dt.scr$PC1, 
               y = dt.scr$PC2, 
               z = dt.scr$PC3, 
               color = as.numeric(dt.scr$trt),
               renderer = "auto",
               pch = dt.scr$sample,
               size = 0.1)
```

# Is sample **02w_CON_1** an outlier?
```{r exclude_sample, fig.height = 8, fig.width = 10}
tmp <- tpm[Geneid %in% levels(tpm$Geneid)[(nrow(tpm) - 8):nrow(tpm)]]
tmp <- melt.data.table(data = tmp,
                       id.vars = 1,
                       measure.vars = 3:ncol(tmp),
                       variable.name = "Sample",
                       value.name = "TPM")
tmp <- merge(dmeta,
             tmp,
             by = "Sample")

p1 <- ggplot(tmp,
             aes(x = Week,
                 y = TPM,
                 colour = Treatment,
                 shape = Replica,
                 group = Treatment)) +
  facet_wrap(~ Geneid,
             scales = "free_y") +
  geom_point(size = 5,
             position = position_dodge(0.5)) +
  ggtitle(gene)
plot(p1)
```
Most abundant molecules in sample **02w_CON_1** are in larger proportions relative the total of the sample (as measured by TPM) compared to all other samples. This could be because less genes were identified in that sample compared to the rest:
```{r total_genes}
colSums(tpm[, -c(1, 2)] > 0)
```

# Conclusion
Exclude sample **02w_CON_1** form the analysis.

# Session Information
```{r info,eval=TRUE}
sessionInfo()
```